cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.api.metadata;
+
+import java.util.Map;
+import java.util.Set;
+import javax.ws.rs.NotFoundException;
+
+/**
+ * The context for reading metadata from program
+ */
+public interface MetadataReaderContext {
+
+  /**
+   * Returns a set of {@link MetadataRecord} representing all metadata (including properties and tags) for the specified
+   * {@link MetadataEntity} in both {@link MetadataScope#USER} and {@link MetadataScope#SYSTEM}.
+   *
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Set<MetadataRecord> getMetadata(MetadataEntity metadataEntity) throws NotFoundException;
+
+  /**
+   * Returns a set of {@link MetadataRecord} representing all metadata (including properties and tags) for the specified
+   * {@link MetadataEntity} in the specified {@link MetadataScope}.
+   *
+   * @throws NotFoundException if the specified entity was not found
```

Comment:
what if the given scope has no metadata, will it throw or return an empty set?

Reply:
Yes, I agree 
https://github.com/caskdata/cdap/pull/9938#discussion_r171979065